src/Drips.sol

```diff
@@ -0,0 +1,501 @@
+        // Skip iterating over `newReceivers` if no new dripping is started
+        uint256 newIdx = newEndTime > _currTimestamp() ? 0 : newReceivers.length;
+        while (true) {
+            // Each iteration gets the next drips update and applies it on the receiver state.
+            // A drips update is composed of two drips receiver configurations,
+            // one current and one new, or from a single drips receiver configuration
+            // if the drips receiver is being added or removed.
+            bool pickCurr = currIdx < currReceivers.length;
+            bool pickNew = newIdx < newReceivers.length;
+            if (!pickCurr && !pickNew) break;
+            if (pickCurr && pickNew) {
+                // There are two candidate drips receiver configurations to create a drips update.
+                // Pick both if they describe the same receiver or the one with a lower address.
+                // The one with a higher address won't be used in this iteration.
+                // Because drips receivers lists are sorted by addresses and deduplicated,
+                // all matching pairs of drips receiver configurations will be found.
+                uint256 currReceiver = currReceivers[currIdx].userId;
+                uint256 newReceiver = newReceivers[newIdx].userId;
+                pickCurr = currReceiver <= newReceiver;
+                pickNew = newReceiver <= currReceiver;
+            }
+            // The drips update parameters
+            uint256 receiver;
+            int128 currAmtPerSec = 0;
+            int128 newAmtPerSec = 0;
+            if (pickCurr) {
+                receiver = currReceivers[currIdx].userId;
+                currAmtPerSec = int128(currReceivers[currIdx].amtPerSec);
+                // Clear the obsolete drips end
+                _setDelta(
+                    s.dripsStates[receiver][assetId].amtDeltas,
+                    cycleSecs,
+                    currEndTime,
+                    currAmtPerSec
+                );
+                currIdx++;
+            }
+            if (pickNew) {
+                receiver = newReceivers[newIdx].userId;
+                newAmtPerSec = int128(newReceivers[newIdx].amtPerSec);
+                // Apply the new drips end
+                _setDelta(
+                    s.dripsStates[receiver][assetId].amtDeltas,
+                    cycleSecs,
+                    newEndTime,
+                    -newAmtPerSec
+                );
+                newIdx++;
+            }
+            // The receiver may have never been used
+            if (!pickCurr) {
+                DripsState storage dripsState = s.dripsStates[receiver][assetId];
+                // The receiver has never been used, initialize it
+                if (dripsState.nextCollectedCycle == 0) {
+                    dripsState.nextCollectedCycle = _currTimestamp() / cycleSecs + 1;
+                }
```

Comment:
I just revisited our nice optimize algorithm here.

I was wondering, why we didn't move the part into an `else` or `else if`.
In the `if(pickCurr)` part?

Lets remind us for the audit to provide some examples for this method. Otherwise it might be hard to understand

Reply:
Because `if(pickCurr) {...} else { <HERE> }` `receiver` isn't set, if `!pickCurr`, the `if(pickNew)` branch is always used and it sets `receiver`.